syntax = "proto3";
package cao_script;

import "cao_common.proto";

message SchemaCard
{
    string name = 1;
    string description = 2;
    repeated string inputs = 3;
    repeated string outputs = 4;
    repeated string constants = 5;
    string ty = 6;
}

message Schema
{
    repeated SchemaCard cards = 1;
}

message CompilationUnit
{
    cao_common.Json encoded = 1;
}

message EntityId
{
    int64 id = 1;
}

message UpdateEntityScriptCommand
{
    cao_common.Uuid userId = 1;
    uint32 entityId = 2;
    cao_common.Uuid scriptId = 3;
}

message UpdateScriptCommand
{
    cao_common.Uuid userId = 1;
    cao_common.Uuid scriptId = 2;
    cao_script.CompilationUnit compilationUnit = 3;
}

message SetDefaultScriptCommand
{
    cao_common.Uuid userId = 1;
    cao_common.Uuid scriptId = 2;
}

/// Empty on success
message CommandResult
{
}

service Scripting
{
    rpc GetBotScriptId(EntityId) returns (cao_common.Uuid) { }
    rpc UpdateEntityScript(UpdateEntityScriptCommand) returns (CommandResult) { }
    rpc UpdateScript(UpdateScriptCommand) returns (CommandResult) { }
    rpc SetDefaultScript(SetDefaultScriptCommand) returns (CommandResult) { }
}
